### ISSUE TITLE
Hanging on when one gpu node return zero as loss in the context of distributed data parallel training

### ISSUE THREAD

**Issue by chenchr** [pytorch/pytorch#23425]
## 🐛 Bug

Hello.
I am training a multi task network on a cluster using 8 gpu with ```Distributed Data Parallel```.
On every iteration, I process one image per gpu.
Suppose I have ```task_a``` and '''task_b``` with ```loss_a``` and ```loss_b``` respectively.
In some case, the ```loss_a``` could be zero on one gpu, which means that there is no loss computed.
Then I just return the zero tensor.
When this situation happens, the code just hang on.

 - PyTorch Version: 1.1.0 
 - OS (e.g., Linux): Ubuntu 
 - How you installed PyTorch (`conda`, `pip`, source): docker conda
 - Python version: 3.6
 - CUDA/cuDNN version:  CUDA 9.0
 - GPU models and configuration: 1080ti
 - Any other relevant information:

**Comment by mrshenli** [comment]
Hi @chenchr, the task has to return the computed output/loss instead of manually create a new zero tensor even if the loss is indeed zero. Because, that loss tensor/variable not only stores the loss value, but also the autograd graph. DDP will use that loss (or output) tensor to find all participating parameters and do the backward reduction. 

If returning the real loss tensor does not solve the problem, please share more details and some code snippet of your trainer.

**Comment by chenchr** [comment]
@mrshenli  Hi, thanks for your reply.
I am training a detection network.
Usually detection's loss consists of ```box regression loss``` and '''classification loss```.
Suppose you have one image that has no box annotation on it, then when calculating the loss, only ```classification loss``` can be conducted, and the ```box regression loss``` has nonsense.  
Therefore, I just return a zero tensor for '''box regression loss'''.
How can I deal with this situation ?
Thanks.

**Comment by mrshenli** [comment]
Hi @chenchr, let me try to better understand the requirement. Are you doing something like the following?

```python
classification_output, box_output = ddp_model(one_image)

if one_image.has_box_annotation:
  box_regression_loss = box_regression_loss_fn(box_output, label)
else:
  box_regression_loss = torch.zeros(...)

classification_loss = classification_loss(classification_output, label)
(classification_loss + box_regression_loss).backward()
```

There are two problems with the above code:

1. Not all DDP outputs contribute to loss: `box_output` was discarded after the forward pass, but DDP backward reduction would still expect gradients for params involved in computing that loss. This will lead to backward hang.
2. If some model params are not used in some iterations, you need to set [find_unused_parameters](https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L196) to true.

The solution to the above problem would be: 1) set `find_unused_parameters`, 2) move the `if-else` clause into `ddp_model`'s `forward` method, and make sure that all ddp outputs contribute to loss.

If your implementation is different from the above, please post some pseudo code to help debugging.

**Comment by chenchr** [comment]
Hi, @mrshenli, exactly it's almost the same, at a little difference.
As the ```box_regression_loss ``` and ```classification_loss ``` is used for statisstics of loss during training.
Therefore, in a network module, I return them as a dict like:
```
def forward(self, x):
    box_regression_loss = ......
    classification_loss = .....
    return_dict = {
        "loss_reg": box_regression_loss,
        "loss_cls": classification_loss ,
    }
    return return_dict

The solution you provide seems to work only when I add these loss into on tensor before return.

**Comment by mrshenli** [comment]
> Maybe I can return scalar zero when no loss_reg computated, and then do some if-else conditioning to deal with the loss reduced from multi gpu.

Yep, doing that (or don't include `box_regression_loss` in the `return_dict` if there is no annotation) together with setting `find_unused_parameters` would work I think.

**Comment by chenchr** [comment]
@mrshenli  A more complicated situation is when I train on multiple gpu using ```distributed data parallel```. The ```loss_reg``` may not be computed on a specific gpu instead of all gpus. 
Then during backward, when the backend begin to sync the graident from all nodes, will this introduces  other problems ? 
Thanks.

**Comment by mrshenli** [comment]
> The loss_reg may not be computed on a specific gpu instead of all gpus.
Then during backward, when the backend begin to sync the gradient from all nodes, will this introduces other problems ?

@chenchr That should be fine, because DDP would still reduce gradients for all params even if there are unused ones on some gpu. The only difference is that when DDP detects some unused params in the forward pass (by walking the autograd graph from the forward output), it would mark those params as backward ready and won't expect their gradients during backward pass.

**Comment by kazimpal87** [comment]
Did this problem ever get solved? I am in a similar situation except that some batches will just have no ground truth and there is no loss. In a single-gpu scenario i can just set the loss to a manually created zero tensor and it works just fine, but in a DDP setting, if one of the gpus gets such a batch, the process hangs indefinitely. Is there any simple way to return a zero loss tensor, with the required gradients (which i guess can also be zero)?

**Comment by zhao-yiqun** [comment]
> Hi @chenchr, let me try to better understand the requirement. Are you doing something like the following?
> 
> ```python
> classification_output, box_output = ddp_model(one_image)
> 
> if one_image.has_box_annotation:
>   box_regression_loss = box_regression_loss_fn(box_output, label)
> else:
>   box_regression_loss = torch.zeros(...)
> 
> classification_loss = classification_loss(classification_output, label)
> (classification_loss + box_regression_loss).backward()
> ```
> 
> There are two problems with the above code:
> 
> 1. Not all DDP outputs contribute to loss: `box_output` was discarded after the forward pass, but DDP backward reduction would still expect gradients for params involved in computing that loss. This will lead to backward hang.
> 2. If some model params are not used in some iterations, you need to set [find_unused_parameters](https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L196) to true.
> 
> The solution to the above problem would be: 1) set `find_unused_parameters`, 2) move the `if-else` clause into `ddp_model`'s `forward` method, and make sure that all ddp outputs contribute to loss.
> 
> If your implementation is different from the above, please post some pseudo code to help debugging.

Hi @mrshenli, I have meet some similar but not the same issue, and also need some advice from you! 
When I use the DDP for training, some iter may have no ground truth for compute the loss. Following the method proposed here, I just skip them in the model.forward(x) and use a loss = torch.zeros(1, required_grad=True) and loss.backward(). while also set find_unused_parameters=True.
In his setting, only part of his network was not updated by loss.backward(). But in my setting, I use the some preprocess method in the forward() to compute whether that is a useful feature after passing it from backbone, If there is no feature, I just use the code like this, And will not continue for the following networks. It means, in some iters of some epochs, there are no parameters to be updated.
`
def forward(x):

    feat = self.backbone(x)

    pairs, nopair = preprocess(feat, x)

    if nopair: # ret here !

         output_dict = {}

         loss_dict.update({‘total_loss’: torch.zeros(1, requires_grad=True)})

         return output_dict, loss_dict

    else: 

         out = self.main_network(feat, x, pairs)

         output_dict.update({‘outputs’: out })

         loss_dict.update({‘total_loss’: self.compute_loss(out, ground_truth)})

        return output_dict, loss_dict

out, loss = model(x)

loss = loss{‘total_loss’}

loss.backward()
`



I’m looking forward to a solution! Thank you very much!

**Comment by zhszysrh** [comment]
> > Hi @chenchr, let me try to better understand the requirement. Are you doing something like the following?
> > ```python
> > classification_output, box_output = ddp_model(one_image)
> > 
> > if one_image.has_box_annotation:
> >   box_regression_loss = box_regression_loss_fn(box_output, label)
> > else:
> >   box_regression_loss = torch.zeros(...)
> > 
> > classification_loss = classification_loss(classification_output, label)
> > (classification_loss + box_regression_loss).backward()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > There are two problems with the above code:
> > 
> > 1. Not all DDP outputs contribute to loss: `box_output` was discarded after the forward pass, but DDP backward reduction would still expect gradients for params involved in computing that loss. This will lead to backward hang.
> > 2. If some model params are not used in some iterations, you need to set [find_unused_parameters](https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L196) to true.
> > 
> > The solution to the above problem would be: 1) set `find_unused_parameters`, 2) move the `if-else` clause into `ddp_model`'s `forward` method, and make sure that all ddp outputs contribute to loss.
> > If your implementation is different from the above, please post some pseudo code to help debugging.
> 
> Hi @mrshenli, I have meet some similar but not the same issue, and also need some advice from you! When I use the DDP for training, some iter may have no ground truth for compute the loss. Following the method proposed here, I just skip them in the model.forward(x) and use a loss = torch.zeros(1, required_grad=True) and loss.backward(). while also set find_unused_parameters=True. In his setting, only part of his network was not updated by loss.backward(). But in my setting, I use the some preprocess method in the forward() to compute whether that is a useful feature after passing it from backbone, If there is no feature, I just use the code like this, And will not continue for the following networks. It means, in some iters of some epochs, there are no parameters to be updated. ` def forward(x):
> 
> ```
> feat = self.backbone(x)
> 
> pairs, nopair = preprocess(feat, x)
> 
> if nopair: # ret here !
> 
>      output_dict = {}
> 
>      loss_dict.update({‘total_loss’: torch.zeros(1, requires_grad=True)})
> 
>      return output_dict, loss_dict
> 
> else: 
> 
>      out = self.main_network(feat, x, pairs)
> 
>      output_dict.update({‘outputs’: out })
> 
>      loss_dict.update({‘total_loss’: self.compute_loss(out, ground_truth)})
> 
>     return output_dict, loss_dict
> ```
> 
> out, loss = model(x)
> 
> loss = loss{‘total_loss’}
> 
> loss.backward() `
> 
> I’m looking forward to a solution! Thank you very much!

I have meet the same issue. how did you solve it in the end?

**Comment by jonaden94** [comment]
> Depending on the exact use-case, there's a rather easy solution to this. Replace this:
> 
> ```python
> if one_image.has_box_annotation:
>   box_regression_loss = box_regression_loss_fn(box_output, label)
> else:
>   box_regression_loss = torch.zeros(...)
> ```
> 
> with this:
> 
> ```python
> if one_image.has_box_annotation:
>   box_regression_loss = box_regression_loss_fn(box_output, label)
> else:
>   box_regression_loss = box_output.sum() * 0.0
> ```
> 
> This way you're effectively returning a zero loss for the sample, but the computation graph is still attached to the loss value so you won't have freezing issues with DDP.
> 
> The general idea is that you just have to ensure that the tensors used to compute the loss scalar are still involved in whatever dummy zero loss value you return. For instance if your model outputs tensors `x` and `y` which are then used to compute the loss, you can use `(x.sum() + y.sum()) * 0.0` as a dummy loss to avoid freezing issues in DDP.

This does not work. The code will still hang. I am using pytorch 1.11.

**Comment by jotaf98** [comment]
DDP requiring a full path between all inputs and outputs across all GPUs seems very brittle.

Why not just allow users to enumerate the Parameters of interest, similar to the Optimizer class?

Then the call graph doesn't matter for synchronization.
